File: server/src/distill_anything/models/event.py
"""Event data models — SQLAlchemy ORM + Pydantic schemas."""

import enum
from datetime import UTC, datetime

from pydantic import BaseModel
from sqlalchemy import Column, DateTime, Enum, Integer, String, Text
from sqlalchemy.orm import DeclarativeBase


class Base(DeclarativeBase):
    pass


class EventType(enum.StrEnum):
    audio = "audio"
    photo = "photo"
    video = "video"
    text = "text"


class UploadStatus(enum.StrEnum):
    received = "received"
    verified = "verified"
    failed = "failed"


class ProcessingStatus(enum.StrEnum):
    pending = "pending"
    processing = "processing"
    completed = "completed"
    failed = "failed"


class Event(Base):
    __tablename__ = "events"

    id = Column(String, primary_key=True)
    type = Column(Enum(EventType), nullable=False)
    timestamp = Column(DateTime, nullable=False)
    raw_file_path = Column(String, nullable=True)
    text_content = Column(Text, nullable=True)
    annotation = Column(Text, nullable=True)
    checksum = Column(String, nullable=True)
    file_size_bytes = Column(Integer, nullable=True)
    upload_status = Column(Enum(UploadStatus), default=UploadStatus.received)
    processing_status = Column(Enum(ProcessingStatus), default=ProcessingStatus.pending)
    transcription = Column(Text, nullable=True)
    description = Column(Text, nullable=True)
    created_at = Column(DateTime, default=lambda: datetime.now(UTC))


class EventCreate(BaseModel):
    """Schema for event creation request metadata."""

    id: str
    type: EventType
    timestamp: datetime
    text_content: str | None = None
    annotation: str | None = None
    checksum: str | None = None
    file_size_bytes: int | None = None


class EventResponse(BaseModel):
    """Schema for event API response."""

    id: str
    type: EventType
    timestamp: datetime
    text_content: str | None = None
    annotation: str | None = None
    checksum: str | None = None
    server_checksum: str | None = None
    file_size_bytes: int | None = None
    upload_status: UploadStatus
    processing_status: ProcessingStatus
    transcription: str | None = None
    description: str | None = None
    created_at: datetime | None = None

    model_config = {"from_attributes": True}

File: client/lib/models/event.dart
import 'package:uuid/uuid.dart';

enum EventType { audio, photo, video, text }

enum UploadStatus { pending, uploading, uploaded, verified, failed }

class Event {
  final String id;
  final EventType type;
  final DateTime timestamp;
  final String? payloadPath;
  final String? textContent;
  final String? annotation;
  final String? checksum;
  final UploadStatus uploadStatus;
  final int? fileSizeBytes;

  Event({
    String? id,
    required this.type,
    DateTime? timestamp,
    this.payloadPath,
    this.textContent,
    this.annotation,
    this.checksum,
    this.uploadStatus = UploadStatus.pending,
    this.fileSizeBytes,
  })  : id = id ?? const Uuid().v4(),
        timestamp = timestamp ?? DateTime.now();

  Map<String, dynamic> toMap() {
    return {
      'id': id,
      'type': type.name,
      'timestamp': timestamp.toIso8601String(),
      'payload_path': payloadPath,
      'text_content': textContent,
      'annotation': annotation,
      'checksum': checksum,
      'upload_status': uploadStatus.name,
      'file_size_bytes': fileSizeBytes,
    };
  }

  factory Event.fromMap(Map<String, dynamic> map) {
    return Event(
      id: map['id'] as String,
      type: _parseEnum(EventType.values, map['type'] as String, EventType.text),
      timestamp: DateTime.parse(map['timestamp'] as String),
      payloadPath: map['payload_path'] as String?,
      textContent: map['text_content'] as String?,
      annotation: map['annotation'] as String?,
      checksum: map['checksum'] as String?,
      uploadStatus: _parseEnum(
          UploadStatus.values, map['upload_status'] as String, UploadStatus.pending),
      fileSizeBytes: map['file_size_bytes'] as int?,
    );
  }

  static T _parseEnum<T extends Enum>(List<T> values, String name, T fallback) {
    for (final v in values) {
      if (v.name == name) return v;
    }
    return fallback;
  }

  Event copyWith({
    String? id,
    EventType? type,
    DateTime? timestamp,
    String? payloadPath,
    String? textContent,
    String? annotation,
    String? checksum,
    UploadStatus? uploadStatus,
    int? fileSizeBytes,
  }) {
    return Event(
      id: id ?? this.id,
      type: type ?? this.type,
      timestamp: timestamp ?? this.timestamp,
      payloadPath: payloadPath ?? this.payloadPath,
      textContent: textContent ?? this.textContent,
      annotation: annotation ?? this.annotation,
      checksum: checksum ?? this.checksum,
      uploadStatus: uploadStatus ?? this.uploadStatus,
      fileSizeBytes: fileSizeBytes ?? this.fileSizeBytes,
    );
  }
}
